config/html-webpack-inline-critical-css-plugin.js

```diff
@@ -0,0 +1,255 @@
+            return Promise.all(styles.map(function (style) {
+              return self.processStyle(style, document);
+            }));
+          })
+          .then(function () {
+            const html = parse5.serialize(document, PARSE5_OPTS);
+            callback(null, { html });
+          })
+          .catch(function (err) {
+            callback(err);
+          });
+      });
+    });
+  }
+
+  processStyle(style, document) {
+    let done = Promise.resolve();
+    let sheet = style.childNodes.length>0 && style.childNodes.map(getNodeValue).join('\n');
+    if (!sheet) return done;
+
+    const ast = css.parse(sheet);
+
+    const before = sheet;
+
+    visit(ast, function (rule) {
+      if (rule.type==='rule') {
+        rule.selectors = rule.selectors.filter(function (sel) {
+          if (sel.match(/:(hover|focus|active)([.[#~&^:*]|\s|\n|$)/)) return false;
+          sel = sel.replace(/::?(?:[a-z-]+)([.[#~&^:*]|\s|\n|$)/gi, '$1');
+          return document.querySelector(sel, document) != null;
+        });
+        // If there are no matched selectors, replace the rule with a comment.
+        if (rule.selectors.length===0) {
+          rule.type = 'comment';
+          rule.comment = '';
+          delete rule.selectors;
+          delete rule.declarations;
+        }
+      }
+
+      if (rule.rules) {
+        // Filter out comments
+        rule.rules = rule.rules.filter(notComment);
+        // If there are no remaining rules, replace the parent rule with a comment.
+        if (rule.rules.length===0) {
+          rule.type = 'comment';
+          rule.comment = '';
+          delete rule.rules;
+        }
+      }
+    });
+
+    sheet = css.stringify(ast, { compress: true });
+
+    if (this.options.minimize || this.options.compress || this.options.minify) {
+      const cssnano = require('cssnano');
+      done = cssnano.process(sheet, {}, { preset: 'default' }).then(function (result) {
+        sheet = result.css;
+      });
+    }
+
+    return done.then(function () {
+      if (sheet.trim().length===0) {
+        // all rules were removed, get rid of the style element entirely
+        sheet.parentNode.removeChild(sheet);
+      }
+      else {
+        // replace the stylesheet inline
+        while (style.lastChild) {
+          style.removeChild(style.lastChild);
+        }
+        style.appendChild(document.createTextNode(sheet));
+      }
+      const name = style.$$name;
+      console.log('\u001b[32mInlined CSS Size' + (name ? (' ('+name+')') : '') + ': ' + (sheet.length / 1000).toPrecision(2) + 'kb (' + ((before.length - sheet.length) / before.length * 100 | 0) + '% of original ' + (before.length / 1000).toPrecision(2) + 'kb)\u001b[39m');
+    });
+  }
+};
+
+
+function visit(node, fn) {
+  if (node.stylesheet) return visit(node.stylesheet, fn);
+
+  node.rules.forEach(function (rule) {
+    if (rule.rules) {
+      visit(rule, fn);
+    }
+    fn(rule);
+  });
+}
+
+
+function readFile(file) {
```

Comment:
WDYT about:

```js
const {promisify} = require('utils');
const fs = require('fs');
const readFile = promisify(fs.readFile);
```

Reply:
yup, if we're doing 8+ that's all this does.